### PR TITLE
Add patches to mirrors 2

### LINKED ISSUE #5344: Patch files not included in mirror
When creating a mirror spack does not seem to include required patch files.

On machine with internet access:
``` 
$ spack mirror create nauty
==> Adding package nauty@2.6r7 to mirror
==> Fetching http://pallini.di.uniroma1.it/nauty26r7.tar.gz
######################################################################## 100.0%
==> nauty@2.6r7 : checksum passed
==> nauty@2.6r7 : added
==> Successfully created mirror in spack-mirror-2017-09-12
  Archive stats:
    0    already present
    1    added
    0    failed to fetch.
```

On machine without internet access:
```
==> Installing nauty
==> Fetching file:///my/local/mirror/spack/mirror/nauty/nauty-2.6r7.tar.gz
######################################################################## 100.0%
==> Staging archive: /my/local/mirror/spack/var/spack/stage/nauty-2.6r7-attii264nx6pydjmpvwpnwzhlca23k5i/nauty26r7.tar.gz
==> Created stage in /my/local/mirror/spack/var/spack/stage/nauty-2.6r7-attii264nx6pydjmpvwpnwzhlca23k5i
==> Fetching file:///my/local/mirror/spack/mirror/nauty/nauty-fix-gt_numorbits.patch
curl: (37) Couldn't open file /my/local/mirror/spack/mirror/nauty/nauty-fix-gt_numorbits.patch
==> Fetching from file:///my/local/mirror/spack/mirror/nauty/nauty-fix-gt_numorbits.patch failed.
==> Fetching https://src.fedoraproject.org/rpms/nauty/raw/0f07d01caf84e9d30cb06b11af4860dd3837636a/f/nauty-fix-gt_numorbits.patch

curl: (22) The requested URL returned error: 403
==> Fetching from https://src.fedoraproject.org/rpms/nauty/raw/0f07d01caf84e9d30cb06b11af4860dd3837636a/f/nauty-fix-gt_numorbits.patch failed.
==> Patch https://src.fedoraproject.org/rpms/nauty/raw/0f07d01caf84e9d30cb06b11af4860dd3837636a/f/nauty-fix-gt_numorbits.patch failed.
==> Error: FetchError: All fetchers failed for spack-stage-e2ZX5W
FetchError: FetchError: All fetchers failed for spack-stage-e2ZX5W

/turquoise/usr/projects/ic/spack/lib/spack/spack/package.py:1029, in do_patch:
     50                            # Touch bad file if anything goes wrong.
     51                            tty.msg('Patch %s failed.' % patch.path_or_url)
     52                            touch(bad_file)
  >> 53                            raise
     54    
     55            if has_patch_fun:
     56                try:
```

### LINKED ISSUE #8637: spack fetch does not fetch patches
### Expected Result

The command "spack fetch [package]" fetches an archive, but not patches required for its installation.
Since some supercomputers block any direct access to the internet, the user must download each patch file manually (in addition to upload prefetched packages). Otherwise, the command "spack install" will freeze, waiting for fetching required patches.
Thus, it would be better to download patch files during a fetch (or to add an option to retrieve them).

### Steps to reproduce the issue

```console
$ spack fetch perl
==> Using cached archive: /panfs/panasas/cnt0041/cvt1605/jrichard/spack/var/spack/cache/perl/perl-5.26.2.tar.gz
==> Using cached archive: /panfs/panasas/cnt0041/cvt1605/jrichard/spack/var/spack/cache/perl/cpanm-5.26.2.tar.gz
$ spack install perl
==> pkgconf is already installed in /panfs/panasas/cnt0041/cvt1605/jrichard/spack/opt/spack/linux-rhel7-x86_64/gcc-6.2.0/pkgconf-1.4.2-ee3vfl4nwkiyh3d4kjxnd35g4k4iiuq2
==> ncurses is already installed in /panfs/panasas/cnt0041/cvt1605/jrichard/spack/opt/spack/linux-rhel7-x86_64/gcc-6.2.0/ncurses-6.1-ebsk7waunygpogrdjpdz66vjejqu3sym
==> readline is already installed in /panfs/panasas/cnt0041/cvt1605/jrichard/spack/opt/spack/linux-rhel7-x86_64/gcc-6.2.0/readline-7.0-be7rllfeh3aoruimktzxv57b5vsesbiv
==> gdbm is already installed in /panfs/panasas/cnt0041/cvt1605/jrichard/spack/opt/spack/linux-rhel7-x86_64/gcc-6.2.0/gdbm-1.14.1-j3hi4vxq5ftk7ablb2wftqo5di23j6tw
==> Installing perl
==> Using cached archive: /panfs/panasas/cnt0041/cvt1605/jrichard/spack/var/spack/cache/perl/perl-5.26.2.tar.gz
==> Using cached archive: /panfs/panasas/cnt0041/cvt1605/jrichard/spack/var/spack/cache/perl/cpanm-5.26.2.tar.gz
==> Staging archive: /panfs/panasas/cnt0041/cvt1605/jrichard/spack/var/spack/stage/perl-5.26.2-nuce4hn4ijcs72zlhd3aaz36kctrxtsc/perl-5.26.2.tar.gz
==> Created stage in /panfs/panasas/cnt0041/cvt1605/jrichard/spack/var/spack/stage/perl-5.26.2-nuce4hn4ijcs72zlhd3aaz36kctrxtsc
==> Staging archive: /panfs/panasas/cnt0041/cvt1605/jrichard/spack/var/spack/stage/resource-cpanm-nuce4hn4ijcs72zlhd3aaz36kctrxtsc/App-cpanminus-1.7042.tar.gz
==> Created stage in /panfs/panasas/cnt0041/cvt1605/jrichard/spack/var/spack/stage/resource-cpanm-nuce4hn4ijcs72zlhd3aaz36kctrxtsc
==> Moving resource stage
	source : /panfs/panasas/cnt0041/cvt1605/jrichard/spack/var/spack/stage/resource-cpanm-nuce4hn4ijcs72zlhd3aaz36kctrxtsc/App-cpanminus-1.7042/
	destination : /panfs/panasas/cnt0041/cvt1605/jrichard/spack/var/spack/stage/perl-5.26.2-nuce4hn4ijcs72zlhd3aaz36kctrxtsc/perl-5.26.2/cpanm/cpanm
==> Fetching https://src.fedoraproject.org/rpms/perl/raw/004cea3a67df42e92ffdf4e9ac36d47a3c6a05a4/f/perl-5.26.1-guard_old_libcrypt_fix.patch
...
```

### LINKED ISSUE #9015: Unexpected behaviour of 'spack mirror create'.
There are several moments with `spack mirror create` that I would like to clarify.
1. The first question is regarding the flag `--one-version-per-spec`. As far a I understand [get_matching_versions](https://github.com/spack/spack/blob/e04fb382b17ecab15ab7dffe3f92a4a282d4a256/lib/spack/spack/mirror.py#L104-L136) recieves a list of concrete specs, which means that there is only one matching version for each spec. So, it seems that we need to pass not only concrete specs but input specs (polished with the information from `packages.yaml`) as well for the flag to make sense. Correct?
2. The second question is related to the first one. The following lines of code seem to be the reason for the flag not to work properly at all:
https://github.com/spack/spack/blob/e04fb382b17ecab15ab7dffe3f92a4a282d4a256/lib/spack/spack/mirror.py#L118-L120
As far as I understand, if we limit the number of output versions, the function stops after checking the first `num_versions` versions. The problem is that not all version among the first `num_versions` potentially satisfy our spec and the condition should be something like
    ```python
    if num_versions and len(matching_spec) >= num_versions:
        break
    ```
    You can try
    ```console
    $ spack mirror create mpich
    ```
3. The third question is regarding the fact that `spack mirror create` does not try to use `list_url`. Compare the results of
    ```console
    $ spack mirror create cdo
    ```
    and
    ```console
    $ spack fetch cdo
    ```
4. The last question is regarding the format of the command line. As far as I understand, the user can specify several specs at once. Specs can contain multivalue variants: `cdo grib2=grib-api`. Does `spack mirror create` parse several such specs correctly? For the command `spack env` we have a special separator `--` for this.

Thank you.

### PR DESCRIPTION
**Pull request by scheibelp**
Anticipated to be ready for review by 9/7 (update 9/5)

Fixes #5344
Fixes #8637 
Fixes #9015 (update 9/5)
See #8585

This is an updated attempt to add patches to mirrors. As of now Spack add the primary source archive and resources to a mirror (e.g. download URLs specified in a Spack `resource` declaration) but does not add patches (e.g. URLs specified with a Spack `patch` declaration). This replaces the custom logic in `mirror.add_single_spec` and downloads all patches and dependency patches.

Like #8699 (which this supercedes):

> This also updates the spack mirror command to add all packages in the concretized DAG (where originally it only added the package specified by the user). This is required in order to collect patches that are specified by dependents: e.g. if X->Y and X requires a patch on Y called Pxy, then Pxy will only be discovered if you create a mirror with X.

Unlike #8699, this makes use of the local Spack cache and any existing download mirrors when creating a new mirror (previously when creating a mirror only the default fetcher was used). This may also resolve issues for packages which have an alternative download link (which is attempted as a backup after the default fetcher) and should resolve errors that are being detected in #8585.

Furthermore, this approach resolves item 3 of https://github.com/spack/spack/issues/9015:

> `spack mirror create` does not try to use list_url

This PR includes edits to address items 1 and 2 of https://github.com/spack/spack/issues/9015:

> 1. The first question is regarding the flag `--one-version-per-spec`. As far a I understand get_matching_versions recieves a list of concrete specs, which means that there is only one matching version for each spec.

This is handled by allowing `mirror.create` to pick additional versions for concrete specs. Currently this just pulls:

* The concrete spec (with whatever version the user asked for)
* After that, the newest versions of the package

While the `mirror.create` library function is capable of dealing with constraints like `foo@2:5`, the `spack mirror create` command always concretizes the spec, so that type of constraint cannot currently be used for example to retrieve versions 2, 3, 4, and 5.

TODOs:

- [x] Make it optional to mirror the entire DAG (this is required to get the dependency patches correct but some users may still want to mirror just the root) (EDIT: this is actually satisfied already with the `-D` option of `spack mirror create`)
- [ ] Currently Spack reports number of hits/misses in terms of the number of packages where some resource was downloaded. This PR updates the statistics to report in terms of the number of individual files downloaded: if 3 packages have a total of 10 resources and patches between them, then it will report adding 10 items (if none of them were present before)
- [x] There was a request in #8699 to optionally download all versions of all packages in a DAG which may also be good to add to this PR - see: https://github.com/spack/spack/pull/8699#issuecomment-404762014

### CODE REVIEW

**Comment by scheibelp** [comment]
@tgamblin this is ready for review

IMO the most controversial aspect is all the monkey-patching in the test. I think some refactoring of fetches etc. is required to improve that aspect substantially so I'd like to handle that later.

**Comment by ax3l** [comment]
Can we keep this PR roll'n? :)

**Comment by healther** [comment]
@scheibelp @tgamblin could we get this merged? I just ran into this again when test driving the current develop

**Comment by IanLee1521** [comment]
@becker33 / @tgamblin -- This is the PR I mentioned that seems like it resolves the issue that we've been discussing.

File: lib/spack/spack/mirror.py [hunk 1]
```diff
@@ -202,57 +213,26 @@ def create(path, specs, **kwargs):
...
-    for spec in version_specs:
-        add_single_spec(spec, mirror_root, categories, **kwargs)
+    mirror_cache = spack.caches.MirrorCache(mirror_root)
+    try:
+        spack.caches.mirror_cache = mirror_cache
+        # Iterate through packages and download all safe tarballs for each
+        for spec in version_specs:
+            add_single_spec(spec, mirror_root, categories, **kwargs)
+    finally:
+        spack.caches.mirror_cache = None
```
Comment (tgamblin) [inline lib/spack/spack/mirror.py]: This try/finally is the part that makes me think that the mirror cache should be moved out of `spack.caches`.  I think it should be integrated somehow with the logic for `FsCache`, but I think factoring all of that properly is a larger task.  There's a lot of global/module-scope state in the fetch and package build logic right now that I would like to get rid of -- this adds to it b/c `mirror_cache` is basically a global here.

File: lib/spack/spack/mirror.py [hunk 2]
```diff
@@ -202,57 +213,26 @@ def create(path, specs, **kwargs):
...
-                    # that to move/copy/create an archive in the mirror.
-                    fetcher.archive(archive_path)
-                    tty.msg("{name} : added".format(name=name))
-
-        if spec_exists_in_mirror:
-            categories['present'].append(spec)
-        else:
-            categories['mirrored'].append(spec)
+        spec.package.do_patch()
+        spec.package.do_clean()
```
Comment (tgamblin) [inline lib/spack/spack/mirror.py]: Oh thank god. `add_single_spec` was a disaster. 👍

File: lib/spack/spack/cmd/mirror.py [hunk 3]
```diff
@@ -63,9 +63,9 @@ def setup_parser(subparser):
         '-D', '--dependencies', action='store_true',
         help="also fetch all dependencies")
     create_parser.add_argument(
-        '-o', '--one-version-per-spec', action='store_const',
-        const=1, default=0,
-        help="only fetch one 'preferred' version per spec, not all known")
+        '-n', '--versions-per-spec', type=int,
+        default=1,
+        help="the number of versions to fetch for each spec")
```
[the remark this answers is not in the document]
Comment (adamjstewart) [reply lib/spack/spack/cmd/mirror.py]: Just make sure to update the Bash completion script next time 😄 
https://github.com/spack/spack/pull/10114/commits/6cea28fb943a6684664b63cb328048a2df83198f

**Pull request merged**